File: app/src/main/java/org/mifos/mobilebanking/ui/adapters/SavingAccountsTransactionListAdapter.java
package org.mifos.mobilebanking.ui.adapters;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.accounts.savings.TransactionType;
import org.mifos.mobilebanking.models.accounts.savings.Transactions;
import org.mifos.mobilebanking.utils.CurrencyUtil;
import org.mifos.mobilebanking.utils.DateHelper;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by dilpreet on 6/3/17.
 */

public class SavingAccountsTransactionListAdapter extends
        RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Transactions> savingAccountsTransactionList;
    private Context context;

    @Inject
    public SavingAccountsTransactionListAdapter() {
        savingAccountsTransactionList = new ArrayList<>();
    }

    public void setContext(Context context) {
        this.context = context;
    }

    public Transactions getItem(int position) {
        return savingAccountsTransactionList.get(position);
    }

    public void setSavingAccountsTransactionList(List<Transactions> savingAccountsTransactionList) {
        this.savingAccountsTransactionList = savingAccountsTransactionList;
        notifyDataSetChanged();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder vh;
        View v = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.row_saving_account_transaction, parent, false);
        vh = new ViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {


        Transactions transaction = getItem(position);

        ((ViewHolder) holder).tvSavingAccountAmount.setText(context.getString(R.string.
                string_and_string, transaction.getCurrency().getDisplaySymbol(), CurrencyUtil.
                formatCurrency(context, transaction.getAmount())));

        ((ViewHolder) holder).tvSavingAccountRunningBalance.setText(context.getString(R.string.
                string_and_string, transaction.getCurrency().getDisplaySymbol(), CurrencyUtil.
                formatCurrency(context, transaction.getRunningBalance())));

        ((ViewHolder) holder).tvTransactionType.setText(transaction.
                getTransactionType().getValue());

        if (transaction.getPaymentDetailData() != null) {
            ((ViewHolder) holder).tvTransactionDetailData.setVisibility(View.VISIBLE);
            ((ViewHolder) holder).tvTransactionDetailData.setText(transaction.
                    getPaymentDetailData().getPaymentType().getName());
        }

        ((ViewHolder) holder).tvTransactionDate.setText(DateHelper.
                getDateAsString(transaction.getDate()));

        ColorSelect color = getColor(transaction.getTransactionType());
        if (color == ColorSelect.RED) {
            ((ViewHolder) holder).vIndicator.setRotation(180);
            ((ViewHolder) holder).vIndicator.setBackgroundDrawable(
                    ContextCompat.getDrawable(context, R.drawable.triangular_red_view));
        } else {
            ((ViewHolder) holder).vIndicator.setRotation(0);
            ((ViewHolder) holder).vIndicator.setBackgroundDrawable(
                    ContextCompat.getDrawable(context, R.drawable.triangular_green_view));
        }

    }

    private enum ColorSelect {
        RED ,
        GREEN
    }

    @Override
    public int getItemCount() {
        return savingAccountsTransactionList.size();
    }

    private ColorSelect getColor(TransactionType transactionType) {
        if (transactionType.getDeposit())
            return ColorSelect.GREEN;
        if (transactionType.getDividendPayout())
            return ColorSelect.RED;
        if (transactionType.getWithdrawal())
            return ColorSelect.RED;
        if (transactionType.getInterestPosting())
            return ColorSelect.GREEN;
        if (transactionType.getFeeDeduction())
            return ColorSelect.RED;
        if (transactionType.getInitiateTransfer())
            return ColorSelect.RED;
        if (transactionType.getApproveTransfer())
            return ColorSelect.RED;
        if (transactionType.getWithdrawTransfer())
            return ColorSelect.RED;
        if (transactionType.getRejectTransfer())
            return ColorSelect.GREEN;
        if (transactionType.getOverdraftFee())
            return ColorSelect.RED;

        return ColorSelect.GREEN;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.tv_transaction_date)
        TextView tvTransactionDate;

        @BindView(R.id.tv_transaction_type)
        TextView tvTransactionType;

        @BindView(R.id.tv_transaction_detail_data)
        TextView tvTransactionDetailData;

        @BindView(R.id.tv_savingAccountAmount)
        TextView tvSavingAccountAmount;

        @BindView(R.id.tv_saving_account_running_balance)
        TextView tvSavingAccountRunningBalance;

        @BindView(R.id.v_indicator)
        View vIndicator;

        public ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
        }
    }
}

File: app/src/main/java/org/mifos/mobile/ui/adapters/SavingAccountsTransactionListAdapter.kt
package org.mifos.mobile.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import org.mifos.mobile.R
import org.mifos.mobile.databinding.RowSavingAccountTransactionBinding
import org.mifos.mobile.models.accounts.savings.TransactionType
import org.mifos.mobile.models.accounts.savings.Transactions
import org.mifos.mobile.utils.CurrencyUtil.formatCurrency
import org.mifos.mobile.utils.DateHelper.getDateAsString
import javax.inject.Inject

/**
 * Created by dilpreet on 6/3/17.
 */
class SavingAccountsTransactionListAdapter @Inject constructor() :
    RecyclerView.Adapter<SavingAccountsTransactionListAdapter.ViewHolder>() {

    private var savingAccountsTransactionList: List<Transactions?>?
    private var context: Context? = null

    init {
        savingAccountsTransactionList = ArrayList()
    }

    fun setContext(context: Context?) {
        this.context = context
    }

    fun getItem(position: Int?): Transactions? {
        return position?.let { savingAccountsTransactionList?.get(it) }
    }

    fun setSavingAccountsTransactionList(savingAccountsTransactionList: List<Transactions?>?) {
        if (savingAccountsTransactionList != null) {
            this.savingAccountsTransactionList =
                savingAccountsTransactionList
        }
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = RowSavingAccountTransactionBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false,
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val transaction = getItem(position)
        holder.bind(transaction)
    }

    private enum class ColorSelect {
        RED, GREEN
    }

    override fun getItemCount(): Int {
        return savingAccountsTransactionList?.size ?: 0
    }

    private fun getColor(transactionType: TransactionType?): ColorSelect {
        if (transactionType?.deposit == true) {
            return ColorSelect.GREEN
        }
        if (transactionType?.dividendPayout == true) {
            return ColorSelect.RED
        }
        if (transactionType?.withdrawal == true) {
            return ColorSelect.RED
        }
        if (transactionType?.interestPosting == true) {
            return ColorSelect.GREEN
        }
        if (transactionType?.feeDeduction == true) {
            return ColorSelect.RED
        }
        if (transactionType?.initiateTransfer == true) {
            return ColorSelect.RED
        }
        if (transactionType?.approveTransfer == true) {
            return ColorSelect.RED
        }
        if (transactionType?.withdrawTransfer == true) {
            return ColorSelect.RED
        }
        if (transactionType?.rejectTransfer == true) {
            return ColorSelect.GREEN
        }
        return if (transactionType?.overdraftFee == true) {
            ColorSelect.RED
        } else {
            ColorSelect.GREEN
        }
    }

    inner class ViewHolder(private val binding: RowSavingAccountTransactionBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(transaction: Transactions?) {
            with(binding) {
                val (_, transactionType, _, _, date, currency, paymentDetailData, amount, runningBalance) = transaction!!
                val displayedSymbol = currency?.displaySymbol ?: currency?.code
                tvSavingAccountAmount.text = context?.getString(
                    R.string.string_and_string,
                    displayedSymbol,
                    formatCurrency(context, amount!!),
                )
                tvSavingAccountRunningBalance.text = context?.getString(
                    R.string.string_and_string,
                    displayedSymbol,
                    formatCurrency(context, runningBalance!!),
                )
                tvTransactionType.text = transactionType?.value

                if (paymentDetailData != null) {
                    tvTransactionDetailData.visibility = View.VISIBLE
                    tvTransactionDetailData.text = paymentDetailData.paymentType.name
                }

                tvTransactionDate.text = getDateAsString(date)
                val color = getColor(transactionType)

                if (color == ColorSelect.RED) {
                    vIndicator.background =
                        ContextCompat.getDrawable(itemView.context, R.drawable.triangular_red_view)
                } else {
                    vIndicator.background = ContextCompat.getDrawable(
                        itemView.context,
                        R.drawable.triangular_green_view,
                    )
                }
            }
        }
    }
}
